gloo/transport/uv/context.h

```diff
@@ -0,0 +1,244 @@
+/**
+ * Copyright (c) 2019-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <gloo/common/memory.h>
+#include <gloo/transport/context.h>
+
+//
+// READ THIS FIRST!
+//
+// This file is a copy of "gloo/transport/tcp/context.h".
+//
+// Any modifications should be made to that file as well. This file is
+// not a top level construct because it assumes the transport supports
+// unbound buffers, which is not yet the case for the ibverbs
+// transport. When that is done, the functionality in this file may be
+// promoted to the top level context class.
+//
+
+namespace gloo {
+namespace transport {
+namespace uv {
+
+// PendingOpCount keeps track of the number of remote pending
+// operations (both pending send and receive operations) for a single
+// slot, across all pairs. This used to be tracked in the pairs
+// themselves, but to support receive-from-any we need a centralized
+// view into all pending send operations. This class facilitaties that
+// centralized view for both pending send and receive operations.
+//
+// The count for a slot is accessed through the transport context
+// object defined further down in this file.
+//
+class PendingOpCount final {
+ public:
+  using count_t = int8_t;
+
+ private:
+  // Counts either pending send or pending recv operations.
+  // Keeps track of number of non-zero entries such that
+  // instances with all zero counts can be cleaned up.
+  class Count final {
+   public:
+    explicit Count(size_t length) : count_(length), nonzero_(0) {}
+
+    bool empty() const {
+      return nonzero_ == 0;
+    }
+
+    count_t get(size_t rank) const {
+      return count_[rank];
+    }
+
+    // Update the count for the specified rank.
+    // If the count for this rank changes from zero to non-zero, it
+    // increments the non-zero counter. If it changes from non-zero to
+    // zero, it decrements the non-zero counter.
+    count_t update(size_t rank, count_t v) {
+      auto cur = count_[rank];
+      if (cur == 0) {
+        cur += v;
+        if (cur != 0) {
+          nonzero_++;
```

Comment:
Just to confirm, this is protected by the lock in `ContextMutator`, right?

Reply:
Yes, correct.